### PR TITLE
Prevent race condition when continuous delivery is enabled

### PR DESCRIPTION
**Pull request by byroot**
### Context

CircleCI is a bit drunk and sometimes send the same status twice in a very short timeframe:

```ruby
=> [{:url=>
  "https://api.github.com/repos/Shopify/starscream/statuses/3c28bfdcc48fae50d0192aa7ce64742c9c89aedc",
 :id=>453068457,
 :state=>"success",
 :description=>"Your tests passed on CircleCI!",
 :target_url=>"https://circleci.com/gh/Shopify/starscream/43380",
 :context=>"ci/circleci",
 :created_at=>2016-02-16 19:19:18 UTC,
 :updated_at=>2016-02-16 19:19:18 UTC,
 ...
, {:url=>
  "https://api.github.com/repos/Shopify/starscream/statuses/3c28bfdcc48fae50d0192aa7ce64742c9c89aedc4742c9c89aedc",
 :id=>453068453,
 :state=>"success",
 :description=>"Your tests passed on CircleCI!",
 :target_url=>"https://circleci.com/gh/Shopify/starscream/43380",
 :context=>"ci/circleci",
 :created_at=>2016-02-16 19:19:18 UTC,
 :updated_at=>2016-02-16 19:19:18 UTC,
 ...
```

In those cases Shipit is subject to a race condition and can trigger two deploys for the same commit.
While there is clearly a problem on CircleCI side, Shipit should not be subject to such race condition.

### Solution

This PR move the code responsible for triggering automatic deploys in a background job, and `BackgroundJob::Unique` prevent any kind of concurrency.

@angelini @rafaelfranca @etiennebarrie @gmalette for review please.

### CODE REVIEW

File: app/models/shipit/stack.rb [hunk 1]
```diff
@@ -80,6 +80,14 @@ def trigger_deploy(until_commit, user, env: nil)
       deploy
     end
 
+    def trigger_continuous_deploy!
+      return unless deployable?
+      if commit = last_deployable_commit
+        return if commit.deployed?
+        trigger_deploy(commit, commit.committer)
```
Comment (angelini) [inline app/models/shipit/stack.rb]: I'm not super familiar with Ruby style, but if this method has a `!` because it modifies internal state, I would assume `trigger_deploy` would also need to be suffixed with a `!`.

Comment (byroot) [inline app/models/shipit/stack.rb]: Yeah right.

Comment (angelini) [inline app/models/shipit/stack.rb]: Why did you remove the `!` instead of adding it to `trigger_deploy`?

Comment (byroot) [inline app/models/shipit/stack.rb]: To stay consistent with the existing code. If I want to change `trigger_deploy` to `trigger_deploy!` I'll do it in another PR / commit.

Comment (etiennebarrie) [inline app/models/shipit/stack.rb]: From our [Ruby style guide](http://shopify.github.io/ruby-style-guide/):
> Never use a ! at the end of a method name if you have no equivalent method without the bang. Methods are expected to change internal object state; you don't need a bang for that. Bangs are to mark a more dangerous version of a method, e.g. save returns a bool in ActiveRecord, whereas save! will throw an exception on failure.

File: test/models/commits_test.rb [hunk 2]
```diff
@@ -65,11 +65,15 @@ class CommitsTest < ActiveSupport::TestCase
     end
 
     test "updating state to success triggers new deploy when stack has continuous deployment" do
+      Hook.destroy_all
```
Comment (etiennebarrie) [inline test/models/commits_test.rb]: Why is this necessary?

File: app/models/shipit/status_group.rb [hunk 3]
```diff
@@ -9,7 +9,7 @@ def initialize(significant_status, visible_statuses)
       @statuses = visible_statuses
     end
 
-    delegate :success?, :state, to: :significant_status
+    delegate :pending?, :success?, :error?, :failure?, :state, to: :significant_status
```
Comment (etiennebarrie) [inline app/models/shipit/status_group.rb]: I don't see new calls to these methods, was it necessary to add the delegation? I'm not against it though.

**Comment by etiennebarrie** [comment]
Would it be useful to add a test for the specific issue that PR solves? Like creating two statuses and checking only one job is enqueued?

File: app/models/shipit/stack.rb (hunk 1 above)
Comment (byroot) [inline app/models/shipit/stack.rb]: > From our Ruby style guide:

That guideline is :hankey:, and it's for internal apps anyway so meh.

File: app/models/shipit/status_group.rb (hunk 3 above)
Comment (byroot) [inline app/models/shipit/status_group.rb]: Yeah, I figured they were missing during my interactive tests. That shouldn't be part of this PR. Removing.

File: test/models/commits_test.rb (hunk 2 above)
Comment (byroot) [inline test/models/commits_test.rb]: It's not anymore. It was to prevent hooks from firing and generating network calls. But since then I changed the test to only run `ContinuousDeliveryJob`, so I'll :fire: that too.

**Comment by byroot** [comment]
> Would it be useful to add a test for the specific issue that PR solves? Like creating two statuses and checking only one job is enqueued?

No 2 would be enqueued, the fix is based on the fact that those jobs are serialized via `BackgroundJob::Unique`. Concurrency tests are very hard and a PITA to maintain (see Semian). So nope, the current tests are enough IMO.

**Comment by gmalette** [comment]
With this kind of check tho, it means it's possible (although very unlikely) that a race condition will happen as a result of a manual deploy + CI checks, given manual deploys are checked in the controller and CI using serialization. Although I don't think it's worth taking that case too seriously, should we instead serialize all deploy creation in the same manner to avoid this?

**Comment by byroot** [comment]
> it's possible (although very unlikely) that a race condition will happen as a result of a manual deploy + CI checks

Indeed.

>  should we instead serialize all deploy creation in the same manner to avoid this?

I though about it. My personal conclusion is:
  -  if continuous deployment is enabled you are not very likely to trigger a deploy manually, and that the window is very short.
  - If we create the deploy record from background, we can't redirect you immediately to the deploy logs page, we have to poll or something for a while.

So IMO it's lots of added complexity for fixing a very unlikely problem. I'll definitely reconsider if it ever happens though.

**Comment by rafaelfranca** [comment]
LGTM

**Comment by byroot** [comment]
Any other concerns or should I :sheep: ?

**Comment by angelini** [comment]
LGTM. We hit this problem again this morning:

double_deploy
